Use the same min-max version range on the client consistently

mattcaswell opened the pull request:
We need to ensure that the min-max version range we use when constructing the ClientHello is the same range we use when we validate the version selected by the ServerHello. Otherwise this may appear as a fallback or downgrade.

Fixes #6964

mattcaswell commented:
Ping @vdukhovni

mattcaswell commented:
I pushed a new commit that fixes a similar problem where there is a protocol version hole on the server side, e.g. the following should work and negotiate TLSv1.1 (current master fails)
````
$ openssl s_client -no_tls1_3
$ openssl s_server -no_tls1_2
````

vdukhovni commented on ssl/statem/statem_lib.c:
```diff
@@ -1956,6 +1952,8 @@ int ssl_choose_client_version(SSL *s, int version, RAW_EXTENSION *extensions)
  * @min_version: The minimum supported version
  * @max_version: The maximum supported version
+ * @real_version: The maximum version the library is capable of regardless of
+ *                what is enabled for this connection
```
This comment could  be more clear. The actual `real_version` is a subtle beast.  It lies just below the lowest **compile-time** hole that lies above at least one **run-time enabled** protocol.  Thus disabling protocols via `SSL_OP_NO_...`at run-time can only **increase** `real_version` by removing all the protocols below some compile-time hole, bringing a new compile-time ceiling into scope.

Fallback adjustments that involves "raising the floor" is not terribly likely, so in practice fallback will have the same `real_version` as what you'd get pre-fallback.

vdukhovni commented on ssl/statem/statem_lib.c:
```diff
@@ -1665,9 +1665,9 @@ static void check_for_downgrade(SSL *s, int vers, DOWNGRADE *dgrd)
-                || ssl_version_supported(s, TLS1_3_VERSION, NULL))) {
+    } else if (!SSL_IS_DTLS(s)
+            && vers < TLS1_2_VERSION
+            && ssl_version_supported(s, TLS1_2_VERSION, NULL)) {
```
Perhaps another place for a comment clarifying the intent.  I think here we're making sure that a server that disables TLS 1.2, creating a hole between 1.1 and 1.3 can still complete handshakes with clients that support 1.2 and below. If we also tested for 1.3 support, we'd signal a 1.1 downgrade and the client would fail, right?  This looks like the scenario David Benjamin was describing... If so, the change makes sense, but substantive comments are needed to highlight a non-trivial decision made here...

vdukhovni commented:
Code looks good, but requires more detailed explanation.  Once that's in place, should be ready to go IMHO.  More reviews welcome.

kroeckx commented:
Does this fix the testsuite error we're getting against boringssl?

vdukhovni commented:
> @kroeckx Does this fix the testsuite error we're getting against boringssl?
Well all the CI tests pass...

kroeckx commented:
Without the extended tests. They fail on master, in the downgrade protection tests.

mattcaswell commented:
> Does this fix the testsuite error we're getting against boringssl?

Unfortunately not. I'm still investigating that error.

mattcaswell commented:
I've pushed a couple of fixup commits that add a comment and clarify an existing one that should hopefully address the feedback received. One of those fixups also changes a reason code in this PR to keep the reason code consistent with behaviour previously (and therefore to avoid some additional BoringSSL external test failures).

I've also analysed the reason why we are seeing BoringSSL external test failures in master at the moment. The reason is rather interesting. Basically we are using a very old version of the BoringSSL test runner which uses an early draft of TLSv1.3 that uses the TLSv1.3 downgrade sentinels by default. We have also recently switched on the downgrade sentinels by default and the mismatched draft/real RFC implementations do no play well together. My more detailed analysis below. The up shot is that I have added an extra commit (with extended tests) to temporarily disable those failing tests until such time as we update the BoringSSL test runner to a more recent version.

### SendFallbackSCSV

In this test the client is OpenSSL and the server is the boring test runner.
The client and server fail to negotiate TLSv1.3 because the test runner is
using an old draft TLSv1.3 version. The server does however add the
TLSv1.3->TLSv1.2 downgrade sentinel in the ServerHello random. Since we
recently turned on checking of the downgrade sentinels on the client side
this causes the connection to fail.

### VersionNegotiationExtension-TLS11

In this test the test runner is the client and OpenSSL is the server. The
test modifies the supported_versions extension sent by the client to only
include TLSv1.1 (and some other spurious versions), even though the client
does actually support TLSv1.2. The server successfully selects TLSv1.1, but
adds the TLSv1.3->TLSv1.1 downgrade sentinel. This behaviour was recently
switched on by default. The test runner then checks the downgrade sentinel
and aborts the connection because it knows that it really supports TLSv1.2.

### VersionNegotiationExtension-TLS1
### VersionNegotiationExtension-SSL3

The same as VersionNegotiationExtension-TLS11 but for TLSv1 and SSLv3.

### ConflictingVersionNegotiation

In this test the client is the test runner, and OpenSSL is the server. The
client offers TLSv1.2 in ClientHello.version, but also adds a
supported_versions extension that only offers TLSv1.1. The
supported_versions extension takes precedence and the server (correctly)
selects TLSv1.1. However it also adds the TLSv1.3->TLSv1.1 downgrade
sentinel. On the client side it knows it actually offered TLSv1.2 and so the
downgrade sentinel check fails.

vdukhovni commented on ssl/statem/statem_lib.c:
```diff
@@ -1665,9 +1665,15 @@ static void check_for_downgrade(SSL *s, int vers, DOWNGRADE *dgrd)
+                * We need to ensure that a server that disables TLSv1.2
+                * (creating a hole between TLSv1.3 and TLSv1.1) can still
+                * complete handshakes with clients that support TLSv1.2 and
+                * below.
```
This will probably suffice, but I would have added: "and therefore we do not enable the sentinel if TLS 1.3 is enabled but TLS 1.2 is not".

vdukhovni commented on ssl/statem/statem_lib.c:
```diff
@@ -1956,6 +1958,8 @@ int ssl_choose_client_version(SSL *s, int version, RAW_EXTENSION *extensions)
  * @min_version: The minimum supported version
  * @max_version: The maximum supported version
+ * @real_max:    The highest version below the lowest compile time version hole
+ *               that lies above at least one run-time enabled protocol.
```
Again, likely sufficient, but perhaps a bit ambiguous, it is the hole and not `real_max` that lies above at least one run-time enabled protocol. The latter may be equal to rather than lie above that run-time enabled protocol. Slightly more verbose comments are good where there's complicated intent.

mattcaswell commented:
Travis was green.

I took another stab at clarifying the comments. Can you reconfirm?

vdukhovni commented:
> @mattcaswell I took another stab at clarifying the comments. Can you reconfirm?

Yes, it will do, but don't feel that you need to be terse.

In Postfix, Wietse writes detailed comments that explain the rationale for various choices in the code, these are not something one can gleam from the code alone, and these comments are not terse, some run for multiple paragraphs. I've tried to do the same.

OpenSSL still needs to get better at this approximation of literate programming.  We sometimes have non-trivial reasons to write the code we write that go beyond the implementation steps to the reasons for the code in the first place.  When these reasons are non-trivial, we should not be stingy with comments. Write more in the comments than in the commit message. I'll try to highlight more such cases as I happen to come across them in reviews, but I hope that we can all try to be less terse with the exposition of the decisions that lead to the code doing what it does.

mattcaswell commented:
Pushed. Thanks.
